File: GitToolBox/src/main/java/zielu/gittoolbox/util/ReschedulingExecutor.java
package zielu.gittoolbox.util;

import com.google.common.collect.ImmutableList;
import com.google.common.util.concurrent.Futures;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.diagnostic.Logger;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.Future;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

public class ReschedulingExecutor implements Disposable {
  private final Logger log = Logger.getInstance(getClass());
  private final ConcurrentMap<String, Future<?>> tasks = new ConcurrentHashMap<>();
  private final AtomicBoolean active = new AtomicBoolean(true);
  private final ScheduledExecutorService executor;
  private final boolean mayInterrupt;

  public ReschedulingExecutor(ScheduledExecutorService executor, boolean mayInterrupt) {
    this.executor = executor;
    this.mayInterrupt = mayInterrupt;
  }

  public Future<?> schedule(String id, Runnable task, long delay, TimeUnit timeUnit) {
    if (active.get()) {
      ScheduledFuture<?> newFuture = executor.schedule(task, delay, timeUnit);
      log.debug("Scheduled ", id, ": ", task);
      Optional.ofNullable(tasks.put(id, newFuture)).ifPresent(oldFuture -> {
        log.debug("Cancelling ", id, " interrupt=", mayInterrupt, ": ", oldFuture);
        oldFuture.cancel(mayInterrupt);
        log.debug("Cancelled ", id, " interrupt=", mayInterrupt, ": ", oldFuture);
      });
      return newFuture;
    } else {
      log.debug("Schedule ", id, " while inactive: ", task);
      return Futures.immediateCancelledFuture();
    }
  }

  @Override
  public void dispose() {
    if (active.compareAndSet(true, false)) {
      ImmutableList<Future<?>> existingTasks = ImmutableList.copyOf(tasks.values());
      tasks.clear();
      existingTasks.forEach(task -> task.cancel(mayInterrupt));
    }
  }
}

File: src/main/kotlin/zielu/gittoolbox/concurrent/ReschedulingExecutor.kt
package zielu.gittoolbox.concurrent

import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.Disposer
import zielu.gittoolbox.GitToolBoxApp
import zielu.intellij.concurrent.ZDisposableRunnableWrapper
import java.util.concurrent.Future
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

internal class ReschedulingExecutor : Disposable {
  private val lock = ReentrantLock()
  private val active = AtomicBoolean(true)
  private val tasks: MutableMap<String, ScheduledFuture<*>> = HashMap()

  fun scheduleOnce(id: String, task: Runnable, delay: Long, timeUnit: TimeUnit): Future<*>? {
    if (!active.get()) {
      return null
    }

    lock.withLock {
      var finalDelayMillis = TimeUnit.MILLISECONDS.convert(delay, timeUnit)
      val oldTask = tasks[id]
      if (oldTask != null) {
        finalDelayMillis = oldTaskExists(oldTask, delay, timeUnit)
        oldTask.cancel(true)
        log.debug("Cancelled $id: $oldTask")
      }
      val wrappedTask = ZDisposableRunnableWrapper(task)
      Disposer.register(this, wrappedTask)
      val scheduled = GitToolBoxApp.getInstance().map {
        it.schedule(wrappedTask, finalDelayMillis, TimeUnit.MILLISECONDS)
      }.orElse(null)
      log.debug("Scheduled $id in $finalDelayMillis ms: $scheduled")
      scheduled?.apply { tasks[id] = this }
      return scheduled
    }
  }

  private fun oldTaskExists(oldTask: ScheduledFuture<*>, delay: Long, timeUnit: TimeUnit): Long {
    val timeLeftMillis = oldTask.getDelay(TimeUnit.MILLISECONDS)
    val delayMillis = TimeUnit.MILLISECONDS.convert(delay, timeUnit)
    return if (timeLeftMillis < delayMillis) {
      if (taskExpired(oldTask)) {
        delayMillis
      } else {
        timeLeftMillis
      }
    } else {
      delayMillis
    }
  }

  private fun taskExpired(task: ScheduledFuture<*>): Boolean {
    return task.isDone || task.isCancelled
  }


  override fun dispose() {
    if (active.compareAndSet(true, false)) {
      lock.withLock {
        tasks.values.forEach { it.cancel(true) }
        tasks.clear()
      }
    }
  }

  private companion object {
    private val log = Logger.getInstance(ReschedulingExecutor::class.java)
  }
}
